server/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java

```diff
@@ -351,6 +352,41 @@ public SortFieldAndFormat build(QueryShardContext context) throws IOException {
         return new SortFieldAndFormat(field, fieldType.docValueFormat(null, null));
     }
 
+    public boolean canRewriteToMatchNone() {
+        return nestedSort == null && (missing == null || "_last".equals(missing));
+    }
+
+    /**
+     * Returns whether some values of the given {@link QueryShardContext#getIndexReader()} are within the
+     * primary sort value provided in the <code>rawBottomSortValues</code>.
+     */
+    public boolean isBottomSortShardDisjoint(QueryShardContext context, Object[] rawBottomSortValues) {
+        if (rawBottomSortValues == null || rawBottomSortValues.length == 0) {
+            return false;
+        }
+
+        if (canRewriteToMatchNone() == false) {
+            return false;
+        }
+        MappedFieldType fieldType = context.fieldMapper(fieldName);
+        if (fieldType == null) {
+            // unmapped
+            return false;
+        }
+        if (fieldType.indexOptions() == IndexOptions.NONE) {
+            return false;
+        }
+        Object minValue = order() == SortOrder.DESC ? rawBottomSortValues[0] : null;
+        Object maxValue = order() == SortOrder.DESC ? null : rawBottomSortValues[0];
+        try {
+            MappedFieldType.Relation relation = fieldType.isFieldWithinQuery(context.getIndexReader(), minValue, maxValue,
+                true, true, null, DEFAULT_DATE_TIME_FORMATTER.toDateMathParser(), context);
```

Comment:
`rawBottomSortValues` stores the implementation-specific values while `isFieldWithinQuery` expects high-level representations of the values. For instance, I believe that this implementation as-is is incorrect for `date_nanos`, which would return a number of nanos since Epoch in `rawBottomSortValues` and then parse a number of millis in `isFieldWithinQuery`. So we should call `DocValueFormat#format` on the min/max values before passing them to `isFieldWithinQuery`.

Reply:
Good catch, I pushed a change to handle this correctly. I had to pass the original docvalues format in the shard search request in order to use the correct date math parser.